ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp

```diff
@@ -435,188 +643,338 @@ string get_login(string username, string password, string cap)
+  defInteger << def;
+  string strDef = defInteger.str();
+
+  tstring tstrStr = widen(str);
+  tstring tstrDef = widen(strDef);
+  tstring tstrWindowCaption = widen(window_get_caption());
 
-  return atof(gs_str_submitted.c_str());
+  wcsncpy(wstrPromptStr, tstrStr.c_str(), 4096);
+  wcsncpy(wstrTextEntry, tstrDef.c_str(), MAX_PATH);
+  wcsncpy(wstrWindowStr, tstrWindowCaption.c_str(), MAX_PATH);
+
+  HideInput = 0;
+  DialogBoxW(enigma::hInstance, MAKEINTRESOURCEW(993), enigma::hWnd, reinterpret_cast<DLGPROC>(InputBoxHookProc));
+
+  static string strResult;
+  strResult = shorten(wstrTextEntry);
+  char *cstrResult = (char *)strResult.c_str();
+
+  return cstrResult ? strtod(cstrResult, NULL) : 0;
+}
+
+double get_passcode(string str, double def)
+{
+  std::ostringstream defInteger;
+  defInteger << def;
+  string strDef = defInteger.str();
+
+  tstring tstrStr = widen(str);
+  tstring tstrDef = widen(strDef);
+  tstring tstrWindowCaption = widen(window_get_caption());
+
+  wcsncpy(wstrPromptStr, tstrStr.c_str(), 4096);
+  wcsncpy(wstrTextEntry, tstrDef.c_str(), MAX_PATH);
+  wcsncpy(wstrWindowStr, tstrWindowCaption.c_str(), MAX_PATH);
+
+  HideInput = 1;
+  DialogBoxW(enigma::hInstance, MAKEINTRESOURCEW(993), enigma::hWnd, reinterpret_cast<DLGPROC>(InputBoxHookProc));
+
+  static string strResult;
+  strResult = shorten(wstrTextEntry);
+  char *cstrResult = (char *)strResult.c_str();
+
+  return cstrResult ? strtod(cstrResult, NULL) : 0;
 }
 
 bool get_string_canceled() {
   return gs_form_canceled;
 }
 
-string get_open_filename(string filter,string filename,string caption)
+string get_open_filename(string filter, string fname)
 {
-  filter.append("||");
-  const unsigned int l = filter.length();
-  for (unsigned int i = 0; i < l; i++)
-    if (filter[i] == '|') filter[i] = 0;
-
-  char fn[MAX_PATH];
-  strcpy(fn, filename.c_str());
-
-  OPENFILENAME ofn;
-  ofn.lStructSize = sizeof(ofn); ofn.hwndOwner = enigma::hWnd; ofn.hInstance = NULL;
-  ofn.lpstrFilter = filter.c_str(); ofn.lpstrCustomFilter = NULL;
-  ofn.nMaxCustFilter = 0; ofn.nFilterIndex = 0;
-  ofn.lpstrFile = fn; ofn.nMaxFile = MAX_PATH;
-  ofn.lpstrFileTitle = NULL; ofn.nMaxFileTitle = 0;
-  ofn.lpstrInitialDir = NULL; ofn.lpstrTitle = caption.length() ? caption.c_str() : NULL;
-  ofn.Flags=OFN_FILEMUSTEXIST | OFN_HIDEREADONLY | OFN_NOCHANGEDIR;
-  ofn.nFileOffset = 0; ofn.nFileExtension = 0;
-  ofn.lpstrDefExt = NULL; ofn.lCustData = 0;
-  ofn.lpfnHook = NULL; ofn.lpTemplateName = 0;
-
-  bool ret = GetOpenFileName(&ofn);
-  return ret == 0 ? "-1" : fn;
-}
-
-string get_save_filename(string filter, string filename, string caption)
+  OPENFILENAMEW ofn;
+
+  string str_filter = filter.append("||");
+
+  tstring tstr_filter = widen(str_filter);
+  replace(tstr_filter.begin(), tstr_filter.end(), '|', '\0');
+  tstring tstr_fname = widen(fname);
+
+  WCHAR wstr_fname[MAX_PATH];
+  wcsncpy(wstr_fname, tstr_fname.c_str(), MAX_PATH);
+
+  ZeroMemory(&ofn, sizeof(ofn));
```

Comment:
This is another big block of logic that appears in four places. Imagine if someone went to fix something in this function. More than half the time, they won't even *think* of the other four places you're doing exactly the same thing, and the functions will slowly drift apart over time.

Reply:
From what I can tell, there is nothing that needs to be fixed with these functions. Unless you are able to point out what is wrong more specifically, I will only be able to assume you think I'm bad at coding therefore you came to the conclusion in all likelihood something was wrong with all four functions, with nothing to base your views on.

Reply:
What I am saying is that you have made it difficult to *check* if there is something wrong by deleting all the functions and writing new ones instead of overwriting the bodies of each function with the new code. This makes the diff utterly unhelpful.

Reply:
So are you suggesting I should redo this pr and paste everything i did into the bodies of each function this time around? What about the hook procedures I added? wouldn't that mess up the diff too? I don't really know how the diff thing works so I just wanted to check before I did this for you.